debug/slides_source/debug.tex

```diff
@@ -0,0 +1,382 @@
+Traceback (most recent call last):
+[...]
+ValueError: Input dimension mis-match.
+    (input[0].shape[0] = 2, input[1].shape[0] = 3)
+Apply node that caused the error:
+     Elemwise{Composite{((i0 + i0) * i1)}}(
+         <TensorType(float64, vector)>,
+         <TensorType(float64, vector)>)
+Toposort index: 0
+Inputs types: [TensorType(float64, vector),
+               TensorType(float64, vector)]
+Inputs shapes: [(2,), (3,)]
+Inputs strides: [(8,), (8,)]
+Inputs values: [array([ 1.,  1.]), array([ 1.,  1.,  1.])]
+Outputs clients: [['output']]
+\end{lstlisting}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{Error message: 2st part}
+HINT: Re-running with most Theano optimization
+disabled could give you a back-traces when this
+node was created. This can be done with by setting
+the Theano flags ``optimizer=fast\_compile''. If that does not
+work, Theano optimizations can be disabled with
+``optimizer=None''.\newline
+HINT: Use the Theano flag ``exception\_verbosity=high''
+for a debugprint of this apply node.
+\end{frame}
+
+
+\begin{frame}[fragile]
+  \frametitle{Error message: Traceback}
+
+\lstset{language=Python,
+        commentstyle=\itshape\color{blue},
+        stringstyle=\color{violet},
+        basicstyle=\footnotesize,
+        xleftmargin=-1em
+        }
+\begin{lstlisting}
+Traceback (most recent call last):
+  File "test.py", line 9, in <module>
+    f(np.ones((2,)), np.ones((3,)))
+  File "/u/bastienf/repos/theano/compile/function_module.py",
+       line 589, in __call__
+    self.fn.thunks[self.fn.position_of_error])
+  File "/u/bastienf/repos/theano/compile/function_module.py",
+       line 579, in __call__
+    outputs = self.fn()
+\end{lstlisting}
+\end{frame}
+
+
+\begin{frame}[fragile]
+  \frametitle{Error message: optimizer=fast\_compile}
+
+\lstset{language=Python,
+        commentstyle=\itshape\color{blue},
+        stringstyle=\color{violet},
+        }
+\begin{lstlisting}
+Backtrace when the node is created:
+  File "test.py", line 7, in <module>
+    z = z * y
+
+\end{lstlisting}
+\end{frame}
+
+\subsection{DebugMode}
+\begin{frame}[fragile]
+  \frametitle{DebugMode}
+    Checks and double-checks everything, extremely slow
+\begin{itemize}
+\item Compare Python, C and GPU implementations.
+\item Compare before and after each optimization.
+\item By default, raise an error on NaN.
+\item Sensitive: so frequently reported errors are ok.
+\end{itemize}
+\end{frame}
+
+\subsection{Breakpoint}
+\begin{frame}[fragile]
+  \frametitle{Breakpoint Op} Identity-like op with the side effect of
+  enforcing a conditional breakpoint, inside a theano function, based
+  on a symbolic scalar condition.
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{Breakpoint Op: optimization}
+  Theano optimization don't know it.
+\begin{itemize}
+  \item Could slow down computation, but not too much.
+  \item Could disable stability optimization
+  \item It is moved to the GPU (it do not force transfer)
+\end{itemize}
```

Comment:
Actually, it would probably be better to say 'Most Theano optimizations don't recognize it' since the optimizations that transfer to the GPU *do* know about it.